### PR TITLE
added time abstraction MV-32

### PR DESCRIPTION
**Pull request by s-gheldd**

### CODE REVIEW

File: src/main/java/org/varroa/util/clock/Clocks.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    public Clocks(@NotNull final Clock startVarroaClock) {
+        backingMap = new ConcurrentHashMap<>();
+        backingMap.put(START_VARROA_CLOCK, startVarroaClock);
+    }
+
+    /**
+     * Set a new clock that can be shared between threads.
+     *
+     * @param identifier a unique identifier.
```
Comment (YannickWeber) [inline src/main/java/org/varroa/util/clock/Clocks.java]: returns the clock corresponding to the given identifier or adds and returns a new system level precision clock if no clock is registered.

File: src/main/java/org/varroa/util/clock/Clocks.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ */
+public class Clocks {
+
+    @NotNull
+    public static final String START_VARROA_CLOCK = "clock.varroa.start";
+
+    @NotNull
+    private final ConcurrentHashMap<String, Clock> backingMap;
+
+    public Clocks(@NotNull final Clock startVarroaClock) {
```
Comment (YannickWeber) [inline src/main/java/org/varroa/util/clock/Clocks.java]: registerClock(@NotNull final String identifier, @NotNull final Clock clock)

File: src/main/java/org/varroa/util/clock/Clocks.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
+package org.varroa.util.clock;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.util.concurrent.ConcurrentHashMap;
+
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/util/clock/Clocks.java]: type: mus -> must

File: src/main/java/org/varroa/util/clock/Clocks.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
+package org.varroa.util.clock;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.util.concurrent.ConcurrentHashMap;
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/util/clock/Clocks.java]: is registered under {@link #START_VARROA_CLOCK} at Varroa start.

File: src/main/java/org/varroa/util/clock/Clocks.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package org.varroa.util.clock;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.time.Clock;
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/util/clock/Clocks.java]: Utility class

File: src/main/java/org/varroa/util/clock/Clocks.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /**
+     * Set a new clock that can be shared between threads.
+     *
+     * @param identifier a unique identifier.
+     * @param clock      the shared clock.
+     */
+    public void setClock(@NotNull final String identifier, @NotNull final Clock clock) {
+        backingMap.put(identifier, clock);
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/util/clock/Clocks.java]: |`line 47`|
|---|
Why do we return a system level precision clock if the clock for the identifier is absent?

Comment (SgtSilvio) [inline src/main/java/org/varroa/util/clock/Clocks.java]: |`line 51`|
|---|
Imho the methods getInstant and getMillis are unnecessary, as they can be achieved by getClock().instant() / getClock().millis()

File: src/test/java/org/varroa/util/clock/ClocksTest.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import org.junit.Test;
+
+import java.time.Clock;
+
+import static org.junit.Assert.assertNotNull;
+
+@SuppressWarnings("NullabilityAnnotations")
+public class ClocksTest {
+
+    @NotNull
```
Comment (SgtSilvio) [inline src/test/java/org/varroa/util/clock/ClocksTest.java]: NotNull is wrong here, remove

File: src/test/java/org/varroa/util/UtilModuleTest.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import org.varroa.util.clock.Clocks;
+
+import java.time.Clock;
+
+import static org.junit.Assert.assertSame;
+
+@SuppressWarnings("NullabilityAnnotations")
+public class UtilModuleTest {
+
+    @NotNull
```
Comment (SgtSilvio) [inline src/test/java/org/varroa/util/UtilModuleTest.java]: NotNulls wrong here, remove them

**Pull request merged**